### PR TITLE
Fix UI language selection in virtualbox

### PR DESCRIPTION
**Pull request by acmelabrat**
Previously only English was available as UI language

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by acmelabrat** [comment]
The fix is mimicked from Debian install of VirtualBox.

Tested this on NixOS by using `nixos-rebuild -I <path_to_local_repo> switch`.

One peculiarity is that virtualbox installed under `environment.systemPackages` shows the language selection correctly (more than built-in/English) without this patch. Mind you, VirtualBox of course can't start any VMs if it is installed this way.

**Review by FriedrichAltheide — CHANGES_REQUESTED** [review]
Thank you very much for the PR. Could you please reword your commit message to "virtualbox: Fix UI language selection in virtualbox"?

**Comment by acmelabrat** [comment]
> Thank you very much for the PR. Could you please reword your commit message to "virtualbox: Fix UI language selection in virtualbox"?

Sure, no problem

**Review by FriedrichAltheide — APPROVED** [review]

**Review by FriedrichAltheide — APPROVED** [review]

**Comment by Yarny0** [comment]
Thanks, @acmelabrat, for fixing this!

Tested after cherry-picking onto [current nixos-unstable](https://github.com/NixOS/nixpkgs/tree/23e89b7da85c3640bbc2173fe04f4bd114342367):

- [x] `virtualboxHardened` builds
- [ ] `nixosTests.virtualbox` all build/pass
- [x] GUI is finally localized!
- [x] very simple VM starts

This is amazing!

However, this breaks the `nixosTests.virtualbox.net-hostonlyif` vm test: It fails after 30 minutes with:

```
command `su - alice -c 'for i in $(seq 1000); do if ipaddr="$(VBoxManage guestproperty get test1 /VirtualBox/GuestInfo/Net/1/V4/IP | sed -n -e '"'"'s/^Value: //p'"'"')" && [ -n "$ipaddr" ]; then echo "$ipaddr"; exit 0; fi; sleep 1; done; echo '"'"'Could not get IPv4 address for test1!'"'"' >&2; exit 1'` failed (exit code 1)
```

From investiagting further, it seems it spends most of its time in `wait_for_vm_boot_test{1,2}`, but I failed to find out why this happens.  I suspect VirtualBox now responds to one of the `VBoxManaged` calls with localized output which isn't recognized by the test script anymore.

**Comment by acmelabrat** [comment]
First off, thank you for your feedback @Yarny0 !

> However, this breaks the `nixosTests.virtualbox.net-hostonlyif` vm test: It fails after 30 minutes with:
> 
> ```
> command `su - alice -c 'for i in $(seq 1000); do if ipaddr="$(VBoxManage guestproperty get test1 /VirtualBox/GuestInfo/Net/1/V4/IP | sed -n -e '"'"'s/^Value: //p'"'"')" && [ -n "$ipaddr" ]; then echo "$ipaddr"; exit 0; fi; sleep 1; done; echo '"'"'Could not get IPv4 address for test1!'"'"' >&2; exit 1'` failed (exit code 1)
> ```
> 
> From investiagting further, it seems it spends most of its time in `wait_for_vm_boot_test{1,2}`, but I failed to find out why this happens. I suspect VirtualBox now responds to one of the `VBoxManaged` calls with localized output which isn't recognized by the test script anymore.

Just to cover all bases: Does this test pass without the cherry-picked commit?

If you're correct and it is because of the localization, this would mean that the test would need to be adjusted.

If it expects files at a certain location, either the files need to be distributed more fine-grained (not seeing forward to that) or the test needs to be changed accordingly.

**Comment by Yarny0** [comment]
tl;dr: All tests pass, I approve this PR without any reservations, I'm sorry for the false alarm.

---

> Does this test pass without the cherry-picked commit?

Yes.  To avoid nix just fetching the test's result from the binary cache, I added an extra empty line somewhere in the test's python code, and it built flawlessly.

Today I tried again, with the exact same tree as yesterday.  Now `simple-gui` failed, also after 15 or more minutes, with a different error: `VirtualBox VM didn't shut down within 2 minutes`.  After adding some debug `print` statements, `simple-gui` passed, but `simple-cli` timed out (didn't note the error this time).  Then I retried from pure `nixos-unstable` and re-ran the tests again (by adding empty lines to the code).  Again, `net-hostonlyif` failed (with the same error as yesterday).

*sigh*

Maybe there is some -- very time consuming and therefore very frustrating -- flakeyness somewhere in the test's logic.  However, after searching for "`nixos.tests.virtualbox`" in https://hydra.nixos.org/jobset/nixos/trunk-combined#tabs-jobs , it seems Hydra is not experiencing that flakeyness.

In any case, I no longer see this PR at fault for any test-related trouble.  Sorry for the confusion, @acmelabrat .  From my point of view, your PR can (and should) be merged!

**Review by 0x4A6F — APPROVED** [review]

**Comment by acmelabrat** [comment]
> tl;dr: All tests pass, I approve this PR without any reservations, I'm sorry for the false alarm.

@Yarny0 , no problem. It's still valuable that you looked into it. I am too novice to know for sure, but it sounds like you might've found an issue with the tests themselves.

**Comment by acmelabrat** [comment]
Is it possible to backport this PR to 24.11 by any chance? :sweat_smile:
